✅ Pull-to-Refresh implementation

ElyasAsmad opened the pull request:
# Description

Refreshes the schedule data from _schedule1.php_

Fixes #50 

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

## Update Preview
![p-t-r iiumschedule](https://user-images.githubusercontent.com/78308067/198261047-c3b10437-16d9-4cd0-94de-0737be6f9b82.gif)

iqfareez commented on lib/views/saved_schedule/saved_schedule_layout.dart:
```diff
@@ -42,6 +61,82 @@ class _SavedScheduleLayoutState extends State<SavedScheduleLayout> {
+        hexColor: colour
+      );
+    }).toList();
+    currentSchedule.lastModified = DateTime.now().toString();
+    widget._box.put(scheduleIdx, currentSchedule);
```
Can be replaced with 
```
currentSchedule.save();
```

iqfareez reviewed (commented):
Salam. Thank you for this very nice addition!! 😄 I always wanted to add this feature but couldn't find time to do it.

Suggestion: Skip update subject where venue is empty. For example, in my cases, my lab venue in imaalum is blank. The lecturer update abt the venue through ms teams etc, so I updated manually in the schedule. With this update, seems like it can override my previously set schedule.

ElyasAsmad commented on lib/views/saved_schedule/saved_schedule_layout.dart:
```diff
@@ -42,6 +61,82 @@ class _SavedScheduleLayoutState extends State<SavedScheduleLayout> {
+       * 
+       * TODO: Implement colour generator class (maybe use Material 3? Refer #30)
+       */
+      int colour = 0xFF000000 | (Random().nextInt(0xFFFFFF));
+      return SavedSubject.fromSubject(
+        subject: subject,
+        subjectName: subject.title,
+        hexColor: colour
+      );
+    }).toList();
```
Hello Iqmal,

Regarding the random colour thing, actually I was having some problem while updating the Hive object, as the new fetched data overrides the previous colour used, causing the app to crash as the colour value is null (refer https://github.com/iqfareez/iium_schedule/issues/50#issuecomment-1293314349).

Anyways, I will try to find a workaround for this matter. 😎👍🏻

iqfareez replied:
Nice works. Seems like the random colour thingy has been resolved., right?

ElyasAsmad replied:
Yep! 

[099032d](https://github.com/iqfareez/iium_schedule/pull/51/commits/099032d1813252fb7a0a78176c3d15c2c8e2b518)

ElyasAsmad commented on lib/views/saved_schedule/saved_schedule_layout.dart:
```diff
@@ -42,6 +61,82 @@ class _SavedScheduleLayoutState extends State<SavedScheduleLayout> {
+    */
+
+    // Get kuliyyah code (e.g: KICT) from tokenized title
+    final kuliyyah = name.split(' ').elementAt(0);
```
I will try to implement the feature to save kuliyyah into HiveDB. New commit will be submitted afterwards.

iqfareez commented on lib/views/saved_schedule/saved_schedule_layout.dart:
```diff
@@ -42,6 +62,94 @@ class _SavedScheduleLayoutState extends State<SavedScheduleLayout> {
+       * 
+       * Refer https://github.com/iqfareez/iium_schedule/pull/51#pullrequestreview-1158961053
+       */
+      if (updatedSubject.venue == '') return prevSubject;
```
```suggestion
      if (updatedSubject.venue == null) return prevSubject;
```
Checking if the string is empty doesn't guarantee the venue is empty. I think checking for null is better, in-line with the albiruni implementation:

https://github.com/iqfareez/albiruni/blob/b8c561eb33ca71625b231bcb3fab993ed64a8952/lib/src/albiruni_base.dart#L152

ElyasAsmad replied:
Fixed the code, thanks for the new knowledge!

iqfareez replied:
No problem

ElyasAsmad commented:
Hello @iqfareez !

In these commits, I have added support for HiveDB to save student's kuliyyah and regenerated Hive's adapter files for this project.

For the #51 feature, I have updated the `onRefresh` method to use the data from HiveDB instead of using the hardcoded kuliyyah information in the schedule title (which is very unsafe).

iqfareez reviewed (approved):
lgtm!

![You've Got LGTM](https://media.giphy.com/media/W7K4C2BnqqafLhHwpX/giphy.gif)

However, before merging this PR, I'm gonna work with #47 first.

iqfareez commented:
@all-contributors add @ElyasAsmad  for code

iqfareez commented:
Hrp jadi. @all-contributors add @ElyasAsmad for code
